src/marlin_pc/constraints.rs

```diff
@@ -0,0 +1,1978 @@
+    data_structures::LabeledCommitment,
+    kzg10::{Proof, VerifierKey as KZG10VerifierKey},
+    marlin_pc::{
+        data_structures::{Commitment, VerifierKey},
+        MarlinKZG10, PreparedCommitment, PreparedVerifierKey,
+    },
+    pc_constraints::{EvaluationsVar, PCCheckVar, QuerySetVar},
+    vec, BTreeMap, BTreeSet, BatchLCProof, LinearCombinationVar, PrepareVar, String, ToString, Vec,
+};
+use ark_ec::{CycleEngine, PairingEngine};
+use ark_ff::{fields::Field, PrimeField, ToConstraintField};
+use ark_nonnative_field::{NonNativeFieldMulResultVar, NonNativeFieldVar};
+use ark_poly::UVPolynomial;
+use ark_r1cs_std::{
+    alloc::{AllocVar, AllocationMode},
+    bits::{boolean::Boolean, uint8::UInt8, ToBitsGadget},
+    eq::EqGadget,
+    fields::{fp::FpVar, FieldVar},
+    groups::CurveVar,
+    pairing::PairingVar,
+    select::CondSelectGadget,
+    R1CSVar, ToBytesGadget, ToConstraintFieldGadget,
+};
+use ark_relations::r1cs::{ConstraintSystemRef, Namespace, Result as R1CSResult, SynthesisError};
+use ark_std::ops::Div;
+use core::{borrow::Borrow, convert::TryInto, marker::PhantomData, ops::MulAssign};
+
+/// Var for the verification key of the Marlin-KZG10 polynomial commitment scheme.
+#[allow(clippy::type_complexity)]
+pub struct VerifierKeyVar<
+    CycleE: CycleEngine,
+    PG: PairingVar<CycleE::E2, <CycleE::E1 as PairingEngine>::Fr>,
+> where
+    <CycleE::E2 as PairingEngine>::G1Projective: MulAssign<<CycleE::E1 as PairingEngine>::Fq>,
+    <CycleE::E2 as PairingEngine>::G2Projective: MulAssign<<CycleE::E1 as PairingEngine>::Fq>,
+    <CycleE::E2 as PairingEngine>::G1Affine:
+        ToConstraintField<<<CycleE::E1 as PairingEngine>::Fr as Field>::BasePrimeField>,
+    <CycleE::E2 as PairingEngine>::G2Affine:
+        ToConstraintField<<<CycleE::E1 as PairingEngine>::Fr as Field>::BasePrimeField>,
+{
+    /// Generator of G1.
+    pub g: PG::G1Var,
+    /// Generator of G2.
+    pub h: PG::G2Var,
+    /// Generator of G1, times first monomial.
+    pub beta_h: PG::G2Var,
+    /// Used for the shift powers associated with different degree bounds.
+    pub degree_bounds_and_shift_powers:
+        Option<Vec<(usize, FpVar<<CycleE::E1 as PairingEngine>::Fr>, PG::G1Var)>>,
+}
+
+impl<CycleE, PG> VerifierKeyVar<CycleE, PG>
+where
+    CycleE: CycleEngine,
+    PG: PairingVar<CycleE::E2, <CycleE::E1 as PairingEngine>::Fr>,
+    <CycleE::E2 as PairingEngine>::G1Projective: MulAssign<<CycleE::E1 as PairingEngine>::Fq>,
+    <CycleE::E2 as PairingEngine>::G2Projective: MulAssign<<CycleE::E1 as PairingEngine>::Fq>,
+    <CycleE::E2 as PairingEngine>::G1Affine:
+        ToConstraintField<<<CycleE::E1 as PairingEngine>::Fr as Field>::BasePrimeField>,
+    <CycleE::E2 as PairingEngine>::G2Affine:
+        ToConstraintField<<<CycleE::E1 as PairingEngine>::Fr as Field>::BasePrimeField>,
+{
+    /// Find the appropriate shift for the degree bound.
+    #[tracing::instrument(target = "r1cs", skip(self, cs))]
+    pub fn get_shift_power(
```

Comment:
Review needed: This is a tricky function that obtains the degree bounds & shift power via PIR.

Reply:
PIR?

Reply:
Hmm not sure I understand the strategy in this method, can you add some comments on the high level structure, as well as comments explaining which parts of the code implement the corresponding parts of the strategy?